parser/parser.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/NeowayLabs/abad/internal/utf16"
+	"github.com/NeowayLabs/abad/lexer"
+	"github.com/NeowayLabs/abad/token"
+)
+
+type (
+	Parser struct {
+		tokens    <-chan lexer.Tokval
+		lookahead []lexer.Tokval
+
+		filename string
+	}
+
+	parserfn func(*Parser) (ast.Node, error)
+)
+
+// used when the tokens is over
+var tokEOF = lexer.Tokval{
+	Type: token.EOF,
+}
+
+var (
+	keywordParsers = map[token.Type]parserfn{}
+	literalParsers = map[token.Type]parserfn{
+		token.Decimal:     parseDecimal,
+		token.Hexadecimal: parseHex,
+	}
+	unaryParsers map[token.Type]parserfn
+)
+
+func init() {
+	unaryParsers = map[token.Type]parserfn{
+		token.Minus: parseUnary,
+		token.Plus:  parseUnary,
+	}
+}
+
+// Parse input source into an AST representation.
+func Parse(fname string, code string) (*ast.Program, error) {
+	p := Parser{
+		tokens:   lexer.Lex(utf16.Encode(code)),
+		filename: fname,
+	}
+
+	return p.parse()
+}
+
+func (p *Parser) parse() (*ast.Program, error) {
+	var nodes []ast.Node
+
+	for {
+		node, eof, err := p.parseNode()
+		if err != nil {
+			return nil, err
+		}
+
+		if eof {
+			break
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return &ast.Program{
+		Nodes: nodes,
+	}, nil
+}
+
+func (p *Parser) parseNode() (n ast.Node, eof bool, err error) {
+	p.scry(1)
+
+	tok := p.lookahead[0]
+	if tok.Type == token.EOF {
+		return nil, true, nil
+	}
+
+	if tok.Type == token.Illegal {
+		_, err := parseIllegal(p)
+		return nil, false, err
+	}
+
+	var parser parserfn
+	var hasparser bool
+
+	for _, parsers := range []map[token.Type]parserfn{
+		keywordParsers,
+		literalParsers,
+		unaryParsers,
+	} {
+		parser, hasparser = parsers[tok.Type]
+		if hasparser {
+			break
+		}
+	}
+
+	if !hasparser {
+		return nil, false, p.errorf(tok, "invalid token: %s", tok)
+	}
+
+	node, err := parser(p)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// parsers should not leave tokens not processed
+	// in the lookahead buffer.
+	if len(p.lookahead) != 0 {
+		panic("parsers not handling lookahead correctly")
+	}
+	return node, false, nil
+}
+
+// next token
+func (p *Parser) next() lexer.Tokval {
+	tok, ok := <-p.tokens
+	if !ok {
+		return tokEOF
+	}
+	return tok
+}
+
+// scry foretell the future using a crystal ball. Amount is how much
+// of the future you want to foresee.
+func (p *Parser) scry(amount int) []lexer.Tokval {
+	if len(p.lookahead)+amount > 2 {
+		panic("lookahead > 2")
+	}
+
+	sz := len(p.lookahead)
+	for i := 0; i < amount-sz; i++ {
+		val := p.next()
+		p.lookahead = append(p.lookahead, val)
+		if val.Type == token.EOF {
+			break
+		}
+	}
+
+	return p.lookahead
+}
+
+// forget what you had foresee
+func (p *Parser) forget(amount int) {
+	for i := 0; i < amount; i++ {
+		p.lookahead = p.lookahead[1:]
+	}
+}
+
+func parseIllegal(p *Parser) (ast.Node, error) {
+	tok := p.lookahead[0]
+	return nil, p.errorf(tok, "invalid token: %s",
+		tok.Value)
+}
+
+func parseDecimal(p *Parser) (ast.Node, error) {
+	tok := p.lookahead[0]
```

Comment:
Is the lookahead used again along the lifetime of the functions ? if not we could have something like a p.pop() that already forgets 1

Reply:
p.next() does that. This way you dont need the lookahead.